### PR TITLE
Implement fork choice rule

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
This PR aims to implement Fork Choice Rule mentioned in issue #214  and partly in #368 

For this PR to be complete, it needs to:

- [x] Implement get_justified_head
- [x] Implement get_justified_head_state
- [x] Implement get_latest_attestation_target
- [x] Implement get_ancestor
- [x] Implement get_latest_attestation_target
- [x] Implement get_children (currently stubbed out)
- [x] Implement lmd_ghost
- [x] Stub out verifyAttestation (specific verification function only to be used in lmd_ghost) 
- [x] Update validatorIndex to LatestAttestation mapping every time we hear a new attestation

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      }
+
+      head =
+          children.stream()
+              .max(
+                  Comparator.comparing(
+                      child_block ->
+                          Math.toIntExact(
+                              get_vote_count(store, child_block, attestation_targets).longValue())))
+              .get();
```
Comment (jrhea) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java:67]: that's gangster

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    return vote_count;
+  }
+
+  /*
+   * Spec pseudo-code:
+   *  Let get_children(store: Store, block: BeaconBlock) -> List[BeaconBlock] returns
+   *  the child blocks of the given block.
+   */
+  // TODO: OPTIMIZE: currently goes through all the values in processedBlockLookup
+  public List<BeaconBlock> get_children(ChainStorageClient store, BeaconBlock block) {
```
Comment (jrhea) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java:94]: ouch...good call on the TODO.  i'm thinking that a good optimization would be a new mapping with key = block_root and value = [childblock1,childblock2,...,childblockN] would be worth it.

**Review by jrhea — APPROVED** [review]
lgtm

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java (hunk 1 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java:67]: haha yep stream API I'm coming for you!!!

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java (hunk 2 above)
Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/LmdGhost.java:94]: I agree. That brings up the question: where do we save that? every time we receive a new block?

**Pull request merged**